### PR TITLE
Every ProviderUser receives Apply / Find open email notification

### PR DESCRIPTION
**Pull request by inulty-dfe**
## Context

Our system should send an email notification to every ProviderUser at the start of the cycle notifying them that the Find Service is open. Also, notify each ProviderUser who is associate with a Provider which has relationships that have not been setup yet.

A bug was found which limited this email to be sent only one in the life of a ProviderUser instead of each cycle.

The bug is in two queries. 

1. Query to fetch the Providers that need organisation permission setup
2. Query to test if the present ProviderUser has had a chaser sent to them about the cycle opening.

we need to scope these queries to the current cycle.

## Changes proposed in this pull request

  The worker used to check if the ProviderUser had EVER got the email
  but now we check that the email hasn't been sent in the current cycle.

  That means it will send the email once for each relevant  ProviderUser every cycle.

## Link to Trello card

[Trello Ticket](https://trello.com/c/zCiKeTjZ/1019-bug-startofcyclenotificationworker-only-ever-sends-email-to-any-provideruser-once-and-not-per-cycle)

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/workers/start_of_cycle_notification_worker.rb [hunk 1]
```diff
@@ -9,7 +9,7 @@ def perform(service = 'find')
 
     providers_scope.limit(fetch_limit).each do |provider|
       provider.provider_users.each do |provider_user|
-        unless ChaserSent.exists?(chased: provider_user, chaser_type: service_open_email)
+        if !ChaserSent.where('created_at > ?', CycleTimetable.send("#{service}_opens")).exists?(chased: provider_user, chaser_type: service_open_email)
```
Comment (elceebee) [inline app/workers/start_of_cycle_notification_worker.rb:12]: This is better than the naive date, nice one

**Review by elceebee — APPROVED** [review]

**Comment by inulty-dfe** [comment]
[Bug: StartOfCycleNotificationWorker only ever sends email to any ProviderUser once and not per cycle](https://trello.com/c/zCiKeTjZ/1019-bug-startofcyclenotificationworker-only-ever-sends-email-to-any-provideruser-once-and-not-per-cycle)

File: app/workers/start_of_cycle_notification_worker.rb (hunk 1 above)
Comment (inulty-dfe) [reply app/workers/start_of_cycle_notification_worker.rb:12]: I think we can improve this with a scope actually.

```ruby
scope :since_service_opened, ->(service) { where('created_at > ?', CycleTimetable.send("#{service}_opens")) }
```

```ruby
ChaserSent.since_service_opened(service).exists?(chased: provider_user, chaser_type: service_open_email)
```

WDYT?

**Review by CatalinVoineag — APPROVED** [review]
👍🏻

**Comment by tomas-stefano** [comment]
How many provider users will we sent this time? (rhetoric question) Is worth checking if we need to batch delivery? Hopefully not but is good to double check :)

**Comment by inulty-dfe** [comment]
> How many provider users will we sent this time? (rhetoric question) Is worth checking if we need to batch delivery? Hopefully not but is good to double check :)

I've added specs to demonstrate how the existing throttling system works. Tell me what you think

**Comment by inulty-dfe** [comment]
I'm going to merge this and if we want, we can create a new ticket to make the throttling better.

**Pull request merged**